services/multiplayer/MultiplayerManager.gd

```diff
@@ -0,0 +1,90 @@
+extends Node
+
+# Maybe we can export theses constants to a global variables project ?
+const DEFAULT_PORT = 28000
+const MAX_PEERS = 4
+
+var peer: ENetMultiplayerPeer
+var player_name := "Pirate"
+
+# Names for remote players in id:name format.
+var players := {}
+
+# Signals to let lobby GUI know what's going on.
+signal player_list_changed()
+signal connection_failed()
+signal connection_succeeded()
+signal game_error(error: int)
+
+
+func _ready() -> void:
+	multiplayer.peer_connected.connect(_peer_connected)
+	multiplayer.peer_disconnected.connect(_peer_disconnected)
+	multiplayer.connected_to_server.connect(_connected_to_server)
+	multiplayer.connection_failed.connect(_connection_failed)
+	multiplayer.server_disconnected.connect(_server_disconnected)
+
+
+func _peer_connected(id: int) -> void:
+	register_player.rpc_id(id, player_name)
+
+
+func _peer_disconnected(id: int) -> void:
+	game_error.emit("Player " + players[id] + " disconnected")
+	unregister_player(id)
+
+
+# Only for clients
+func _connected_to_server() -> void:
+	connection_succeeded.emit()
+
+
+# Only for clients
+func _server_disconnected() -> void:
+	game_error.emit("Server disconnected")
+	
+
+# Only for clients
+func _connection_failed() -> void:
+	multiplayer.set_network_peer(null) # Remove peer
+	connection_failed.emit()
```

Comment:
If you want to group functions for client you can use `region` instead of comment:
```suggestion
#region client-only
func _connected_to_server() -> void:
	connection_succeeded.emit()


func _server_disconnected() -> void:
	game_error.emit("Server disconnected")
	

func _connection_failed() -> void:
	multiplayer.set_network_peer(null) # Remove peer
	connection_failed.emit()
#endregion client-only
```